// ts-dde/dist/utils/get-server-invoke.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var edge = require('edge-js');
var modelPath = path.join(path.dirname(__filename), '../../dll');
function getServerInvokerFunc() {
    return edge.func({
        source: modelPath + '/server.cs',
        references: [modelPath + '/NDde.dll'],
        typeName: 'NodeDde.Server',
        methodName: 'GetInvoker',
    });
}
function getServerInvoke(ddeServ) {
    var invokerFunc = getServerInvokerFunc();
    return invokerFunc({
        service: ddeServ.serviceName,
        callbacks: {
            OnBeforeConnect: function (ddeData, cb) {
                ddeServ.emit('before connect', ddeData.topic);
                cb(null, ddeServ.onBeforeConnect(ddeData.topic));
            },
            OnAfterConnect: function (ddeData) {
                ddeServ.emit('after connect', ddeData.service, ddeData.topic);
                ddeServ.onAfterConnect(ddeData.service, ddeData.topic);
            },
            OnDisconnect: function (ddeData) {
                ddeServ.emit('disconnect', ddeData.service, ddeData.topic);
                ddeServ.onDisconnect(ddeData.service, ddeData.topic);
            },
            OnStartAdvise: function (ddeData, cb) {
                ddeServ.emit('start advise', ddeData.service, ddeData.topic, ddeData.item, ddeData.format);
                cb(null, ddeServ.onStartAdvise(ddeData.service, ddeData.topic, ddeData.item, ddeData.format));
            },
            OnStopAdvise: function (ddeData) {
                ddeServ.emit('stop advise', ddeData.service, ddeData.topic, ddeData.item);
                ddeServ.onStopAdvise(ddeData.service, ddeData.topic, ddeData.item);
            },
            OnExecute: function (ddeData, cb) {
                ddeServ.emit('execute', ddeData.service, ddeData.topic, ddeData.command);
                cb(null, ddeServ.onExecute(ddeData.service, ddeData.topic, ddeData.command));
            },
            OnPoke: function (ddeData, cb) {
                ddeServ.emit('poke', ddeData.service, ddeData.topic, ddeData.item, ddeData.data, ddeData.format);
                cb(null, ddeServ.onPoke(ddeData.service, ddeData.topic, ddeData.item, ddeData.data, ddeData.format));
            },
            OnRequest: function (ddeData, cb) {
                ddeServ.emit('request', ddeData.service, ddeData.topic, ddeData.item, ddeData.format);
                cb(null, ddeServ.onRequest(ddeData.service, ddeData.topic, ddeData.item, ddeData.format));
            },
            OnAdvise: function (ddeData, cb) {
                ddeServ.emit('advise', ddeData.topic, ddeData.item, ddeData.format);
                cb(null, ddeServ.onAdvise(ddeData.topic, ddeData.item, ddeData.format));
            },
        },
    }, true);
}
exports.getServerInvoke = getServerInvoke;
